File: src/Day01.java
import java.io.IOException;
import java.math.BigInteger;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Day01 {

    public static void main(String[] args) throws IOException {
        if (args.length == 0) {
            List<String> testInput = readInput("Day01_test");
            assert part1(testInput) == 11;

            List<String> input = readInput("Day01");
            System.out.println(part1(input));
            System.out.println(part2(input));
        } else {
            List<String> lines = Files.lines(Paths.get(args[0])).toList();
            System.out.println(part1(lines));
            System.out.println(part2(lines));
        }
    }

    public static List<int[]> parseDay1(List<String> input) {
        return input.stream()
                .map(line -> line.split(" "))
                .map(it -> Stream.of(it).filter(x -> !x.isBlank()).map(Integer::parseInt).toList())
                .map(parts -> new int[]{parts.getFirst(), parts.get(1)})
                .collect(Collectors.toList());
    }

    public static int part1(List<String> input) {
        List<int[]> pairs = parseDay1(input);
        List<Integer> sortedLeft = pairs.stream().map(p -> p[0]).sorted().toList();
        List<Integer> sortedRight = pairs.stream().map(p -> p[1]).sorted().toList();
        var result = 0;
        for (int i = 0; i < sortedLeft.size(); i++) {
            var l = sortedLeft.get(i);
            var r = sortedRight.get(i);
            result += Math.abs(l - r);
        }
        return result;
    }

    public static long part2(List<String> input) {
        List<int[]> pairs = parseDay1(input);
        List<Integer> lefts = pairs.stream().map(p -> p[0]).sorted().toList();
        Map<Integer, Long> counts = pairs.stream().map(p -> p[1]).collect(Collectors.groupingBy(e -> e, Collectors.counting()));

        long sum = 0;
        for (Integer left : lefts) {
            sum += left * counts.getOrDefault(left, 0L);
        }
        return sum;
    }

    public static List<String> readInput(String name) throws IOException {
        return Files.readAllLines(Paths.get("src/" + name + ".txt"));
    }
}

File: src/Day01.kt
import java.math.BigInteger
import java.security.MessageDigest
import kotlin.io.path.Path
import kotlin.io.path.readText
import kotlin.math.abs

fun main(args:Array<String>) {
    fun parseDay1(input: List<String>): List<Pair<Int, Int>> {
        val pairs = input
            .map { it.split(" ").filter { it.isNotBlank() } }
            .map { (it[0].toInt()) to (it[1].toInt()) }
        return pairs
    }

    fun part1(input: List<String>): Int {
        val pairs = parseDay1(input)
        val sortedLeft = pairs.map { it.first }.sorted()
        val sortedRight = pairs.map { it.second }.sorted()
        return sortedLeft.zip(sortedRight).sumOf { (l, r) ->
            abs(l - r)
        }
    }

    fun part2(input: List<String>): Int {
        val pairs = parseDay1(input)
        val lefts = pairs.map { it.first }.sorted()
        val counts = pairs.map { it.second }.groupingBy { it }.eachCount()
//        val leftCounts = sortedRight.groupingBy { it }.eachCount()
        return lefts.sumOf { it * (counts[it] ?: 0) }
    }

    // Test if implementation meets criteria from the description, like:
//    check(part1(listOf("test_input")) == 1)

    // Or read a large test input from the `src/Day01_test.txt` file:
    if(args.isEmpty()){
        val testInput = readInput("Day01_test")
        check(part1(testInput) == 11)

        // Read the input from the `src/Day01.txt` file.
        val input = readInput("Day01")
        part1(input).println()
        part2(input).println()
    } else {
        val input = Path(args[0]).readText().trim().lines()
        part1(input).println()
        part2(input).println()
    }
}

/**
 * Reads lines from the given input txt file.
 */
fun readInput(name: String) = Path("src/$name.txt").readText().trim().lines()

/**
 * Converts string to md5 hash.
 */
fun String.md5() = BigInteger(1, MessageDigest.getInstance("MD5").digest(toByteArray()))
    .toString(16)
    .padStart(32, '0')

/**
 * The cleaner shorthand for printing output.
 */
fun Any?.println() = println(this)
